Fix in-place collect not reallocating when necessary

the8472 opened the pull request:
Regression introduced in https://github.com/rust-lang/rust/pull/110353.
This was [caught by miri](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri/topic/Cron.20Job.20Failure.20.28miri-test-libstd.2C.202023-11.29/near/404764617)

r? @saethlin

the8472 committed:
Fix in-place collect not reallocating when necessary

the8472 commented on library/alloc/src/vec/in_place_collect.rs:
```diff
@@ -168,7 +168,7 @@ const fn in_place_collectible<DEST, SRC>(
     step_expand: Option<NonZeroUsize>,
 ) -> bool {
-    if DEST::IS_ZST || mem::align_of::<SRC>() < mem::align_of::<DEST>() {
+    if const { SRC::IS_ZST || DEST::IS_ZST || mem::align_of::<SRC>() < mem::align_of::<DEST>() } {
```
This is not part of the fix. the condition was already enforced before, I'm just making it more explicit for readability.

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/7038701014/job/19156175546) [(plain)](https://github.com/rust-lang/rust/commit/4268a02eaa005fc59a2aea4feecfe378fc52817c/checks/19156175546/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

cjgillot commented on library/alloc/src/vec/in_place_collect.rs:
```diff
@@ -186,6 +186,32 @@ const fn in_place_collectible<DEST, SRC>(
+        // dest_sz can't actually be 0 since in_place_collectible() checks for that but const eval doesn't know that
+        let dest_sz = mem::size_of::<DEST>();
+        let src_not_multiple_of_dest =
+            dest_sz == 0 || mem::size_of::<SRC>() % mem::size_of::<DEST>() != 0;
```
Can this const block use src_sz and dest_sz more consistently?

the8472 replied:
Done

asquared31415 commented on library/alloc/src/vec/in_place_collect.rs:
```diff
@@ -186,6 +186,32 @@ const fn in_place_collectible<DEST, SRC>(
+        // dest_sz can't actually be 0 since in_place_collectible() checks for that but const eval doesn't know that
+        let dest_sz = mem::size_of::<DEST>();
+        let src_not_multiple_of_dest =
+            dest_sz == 0 || mem::size_of::<SRC>() % mem::size_of::<DEST>() != 0;
```
Why does this have the `dest_sz == 0` check at all if it's not possible to be 0? IMO it would be more desirable for const eval to panic in case the mentioned check gets removed by mistake (especially since it's not local) and would be clearer to not have that code path if it's meant to be a constant `false || ...`.

the8472 replied:
because the code path still gets instantiated. it just becomes unreachable due to an earlier branch on `in_place_collectible()`, which means the consts still get evaluated. If I panic here it panics at compile time even though the branch is dead.

RalfJung commented:
I can confirm that with this patch applied, the test suite passes in Miri.

saethlin commented:
The current implementation in this PR looks like it is trying to enumerate the cases where reallocation is not permitted, then if none of those cases are hit we fall through to permitting allocation reuse. I would much prefer to see this implemented with the inverse strategy; try to enumerate the cases where allocation reuse _is_ permitted, then fall through to requiring reallocation. Alignment is trivial though, so I don't mind handling that differently.

So I'd prefer to see something where either we early-return `false` and default to returning `true`.

That being said, I slapped together this alternative implementation and it seems to pass the test suite:
```rust

const fn needs_realloc<SRC, DEST>(src_cap: usize, dst_cap: usize) -> bool {
    // Alignment must match exactly. Beyond this check we do not need to reason about it.
    if const { mem::align_of::<SRC>() != mem::align_of::<DEST>() } {
        return true;
    }

    // Exact allocation size matches can be reused
    let src_bytes = src_cap * mem::size_of::<SRC>();
    let dest_bytes = dst_cap * mem::size_of::<DEST>();
    if src_bytes == dest_bytes {
        return false;
    }

    // Everything else needs a realloc
    true
}
```
This implementation seems much simpler to me and I'd prefer to see it (or something similar) used if it is equivalent. Is it? If it isn't, I think we're missing a test.

the8472 commented:
The difference is that I'm doing more `const` checks to turn find cases where we don't need to do runtime checks to figure out whether realloc is needed. I'd have to write codegen tests to catch that.
But sure, I can invert the condition.

the8472 commented:
I rewrote the `needs_realloc` conditions and also added a check for zero-sized allocations to catch up with #113113 which happened after I originally wrote the realloc code.

saethlin commented on library/alloc/src/vec/in_place_collect.rs:
```diff
@@ -186,6 +186,26 @@ const fn in_place_collectible<DEST, SRC>(
+        return src_cap > 0;
+    }
+
+    // If src size is an integer multiple of the destination size
+    // the resulting capacity will also always be an integer multiple without remainder
```
```suggestion
    // If src size is an integer multiple of the destination size, we know the capacity we calculate
    // with src_cap * size_of::<SRC> / size_of::<DEST> does not have a remainder.
```
It took me a long time to understand this comment. It's referring to the division done outside this function. I've tried to make that clear.

r=me with this suggestion or equivalent clarification

the8472 commented:
@bors r=saethlin

bors commented:
:pushpin: Commit 13a843ebcbe536257a8442bf5c26b227d1c2f7c9 has been approved by `saethlin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=saethlin 13a843ebcbe536257a8442bf5c26b227d1c2f7c9 -->
<!-- homu: {"type":"Approved","sha":"13a843ebcbe536257a8442bf5c26b227d1c2f7c9","approver":"saethlin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

RalfJung commented:
@bors p=1
this unblocks miri-test-libstd

bors commented:
:hourglass: Testing commit 13a843ebcbe536257a8442bf5c26b227d1c2f7c9 with merge 15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1...
<!-- homu: {"type":"BuildStarted","head_sha":"13a843ebcbe536257a8442bf5c26b227d1c2f7c9","merge_sha":"15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/7112165906/job/19365856222)
Approved by: saethlin
Pushing 15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"saethlin","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/7112165906/job/19365856222"},"merge_sha":"15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1"} -->

rust-timer commented:
Finished benchmarking commit (15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1): [comparison URL](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.4% | [0.4%, 0.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | 1.0% | [1.0%, 1.0%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.4% | [0.4%, 0.4%] | 1     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.2% | [3.2%, 3.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=17a520aa3769ebbf6653eef8369ab3c459b887aa&end=15bb3e204ad5fbaf53b2b26e47247f3cbbba63f1&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.0%  | [0.0%, 0.2%]   | 26    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.2% | [-0.2%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.0%  | [-0.2%, 0.2%]  | 28    |
</details>

**Bootstrap**: 674.58s -> 675.083s (0.07%)
**Artifact size**: 314.17 MiB -> 314.18 MiB (0.00%)
